### PR TITLE
Assignment9+10 multithreading alan and grant

### PR DESCRIPTION
**Pull request by arose-13**
Alan Rose (@arose-13) and Grant Howard (@Howard404)

### CODE REVIEW

**Comment by zlawson-ewu** [comment]
I will review this.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -59,61 +59,124 @@ public void RunTaskAsync_Success()
...
-        // Pseudo Code - don't trust it!!!
-        string[] hostNames = new string[] { "localhost", "localhost", "localhost", "localhost" };
-        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length*hostNames.Length;
+        string[] hostNames = new string[] { "localhost", "localhost", "localhost", "localhost", "localhost" };
         PingResult result = await Sut.RunAsync(hostNames);
+
+        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length * hostNames.Length;
+        int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
+
+        Assert.AreEqual(expectedLineCount, lineCount);
```
Comment (zlawson-ewu) [inline Assignment/Assignment.Tests/PingProcessTests.cs:126]: Use generic AreEqual<T> here.
```suggestion
        Assert.AreEqual<int>(expectedLineCount, lineCount);
```

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 2]
```diff
@@ -59,61 +59,124 @@ public void RunTaskAsync_Success()
...
+    async public Task RunAsync_IEnumerableHostAddresses_CorrectLineCount()
+    {
+        IEnumerable<string> hostNames = new List<string> { "localhost", "localhost", "localhost", "localhost", "localhost" };
+        CancellationTokenSource tokenSource = new();
+        PingResult result = await Sut.RunAsync(hostNames, tokenSource.Token);
+
+        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length * hostNames.Count();
         int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
+
         Assert.AreEqual(expectedLineCount, lineCount);
```
Comment (zlawson-ewu) [inline Assignment/Assignment.Tests/PingProcessTests.cs:139]:
```suggestion
        Assert.AreEqual<int>(expectedLineCount, lineCount);
```

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 3]
```diff
@@ -59,61 +59,124 @@ public void RunTaskAsync_Success()
...
+        }
+    }
+
     [TestMethod]
 #pragma warning disable CS1998 // Remove this
     async public Task RunLongRunningAsync_UsingTpl_Success()
     {
-        PingResult result = default;
-        // Test Sut.RunLongRunningAsync("localhost");
+        PingResult result = Sut.RunLongRunningAsync("localhost").Result;
```
Comment (zlawson-ewu) [inline Assignment/Assignment.Tests/PingProcessTests.cs:179]: Test shows compile error if pragma statements removed. Remove .Result from method call and then the pragma statements can be removed. Test will then compile and pass.
```suggestion
        PingResult result = Sut.RunLongRunningAsync("localhost");
```

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 4]
```diff
@@ -125,7 +188,7 @@ public void StringBuilderAppendLine_InParallel_IsNotThreadSafe()
         System.Text.StringBuilder stringBuilder = new();
         numbers.AsParallel().ForAll(item => stringBuilder.AppendLine(""));
         int lineCount = stringBuilder.ToString().Split(Environment.NewLine).Length;
-        Assert.AreNotEqual(lineCount, numbers.Count()+1);
+        Assert.AreNotEqual(lineCount, numbers.Count() + 1);
```
Comment (zlawson-ewu) [inline Assignment/Assignment.Tests/PingProcessTests.cs:191]:
```suggestion
        Assert.AreNotEqual<int>(lineCount, numbers.Count() + 1);
```

**Review by zlawson-ewu — COMMENTED** [review]
## Assignment 9&10

## Instructions

1. Implement `PingProcess`' `public Task<PingResult> RunTaskAsync(string hostNameOrAddress)` ✔
   - First implement `public void RunTaskAsync_Success()` test method to test `PingProcess.RunTaskAsync()` using `"localhost"`. ✔
   - Do NOT use async/await in this implementation. ✔
2. Implement `PingProcess`' `async public Task<PingResult> RunAsync(string hostNameOrAddress)` ✔
   - First implement the `public void RunAsync_UsingTaskReturn_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` **without** using async/await. ✔
   - Also implement the `async public Task RunAsync_UsingTpl_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` but this time **DO** using async/await. ✔
3. Add support for an optional cancellation token to the `PingProcess.RunAsync()` signature. ✔
   Inside the `PingProcess.RunAsync()` invoke the token's `ThrowIfCancellationRequested()` method so an exception is thrown. ✔
   Test that, when cancelled from the test method, the exception thrown is an `AggregateException` ✔ with a `TaskCanceledException` inner exception ✔ using the test methods `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping` ✔and `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException` ✔ respectively.
4. Complete/fix **AND test** `async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)` which executes ping for an array of hostNameOrAddresses (which can all be "localhost") **in parallel**, adding synchronization if needed. ✔
   NOTE:
      - The order of the items in the stdOutput is irrelevant and expected to be intermingled.
      - StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
5. Implement **AND test** `public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)` using `Task.Factory.StartNew()` and invoking `RunProcessInternal` with a `TaskCreation` value of `TaskCreationOptions.LongRunning` and a `TaskScheduler` value of `TaskScheduler.Current`. Returning a `Task<PingResult>` is also okay. ✔
   NOTE: This method does **NOT** use `Task.Run`.

## Extra Credit

- Test and implement `PingProcess.RunAsync(System.IProgress<T> progess)` so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  (❌?✔)
- All of the above should be unit tested (❌?✔)
- Choose simplicity over complexity ✔

## Commentary
Overall, nice work! I left line comments for the things I noticed, but these are minor and easily fixed.

**Review by eroadhouse — COMMENTED** [review]
Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
Do NOT use async/await in this implementation. ✔
Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔
Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔ with a TaskCanceledException inner exception ✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.
Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for an array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
The order of the items in the stdOutput is irrelevant and expected to be intermingled.
StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.

Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Assignment/Assignment.Tests/PingProcessTests.cs (hunk 1 above)
Comment (arose-13) [reply Assignment/Assignment.Tests/PingProcessTests.cs:126]: Ah! Missed that! We'll fix that.

**Pull request closed without merging**